t/03-proxy_wasm/hfuncs/metrics/005-record_metric_edge_cases.t

```diff
@@ -0,0 +1,74 @@
+# vim:set ft= ts=4 sts=4 sw=4 et fdm=marker:
+
+use strict;
+use lib '.';
+use t::TestWasmX;
+
+skip_hup();
+no_shuffle();
+
+plan_tests(4);
+run_tests();
+
+__DATA__
+
+=== TEST 1: proxy_wasm metrics - record_metric() no memory, can't expand histogram
+In SIGHUP mode, this test fails if executed after a test that defined more
+metrics than it's possible to fit in `5m`.
+
+--- valgrind
+--- load_nginx_modules: ngx_http_echo_module
+--- main_config eval
+qq{
+    wasm {
+        module hostcalls $ENV{TEST_NGINX_CRATES_DIR}/hostcalls.wasm;
+
+        metrics {
+            slab_size 5m;
+            max_metric_name_length 128;
+        }
+    }
+}
+--- config
+    location /t {
+        proxy_wasm hostcalls 'on_configure=define_metrics \
+                              on=request_headers \
+                              test=/t/metrics/record_histograms \
+                              metrics=h16256 \
+                              value=1 \
+                              metrics_name_len=115';
+
+        proxy_wasm hostcalls 'on_configure=define_metrics \
+                              on=request_headers \
+                              test=/t/metrics/record_histograms \
+                              metrics=h16256 \
+                              value=2 \
+                              metrics_name_len=115';
+
+        proxy_wasm hostcalls 'on_configure=define_metrics \
+                              on=request_headers \
+                              test=/t/metrics/record_histograms \
+                              metrics=h16256 \
+                              value=4 \
+                              metrics_name_len=115';
+
+        proxy_wasm hostcalls 'on_configure=define_metrics \
+                              on=request_headers \
+                              test=/t/metrics/record_histograms \
+                              metrics=h16256 \
+                              value=8 \
+                              metrics_name_len=115';
+
+        proxy_wasm hostcalls 'on_configure=define_metrics \
+                              on=request_headers \
+                              test=/t/metrics/record_histograms \
+                              metrics=h16256 \
+                              value=16 \
+                              metrics_name_len=115';
+        echo ok;
+    }
+--- error_log
+cannot expand histogram
```

Comment:
We should check all error log level from error logs initiated from our C code. And also ensure that all produced error logs, whether low-level in our metrics ABI or host function traps etc, are all documented in failure tests. Does this failure also create a trap? It seems so as `[crit]` is not included below and if so, then we should test it too.

Reply:
Good point.

This specific test is not a failure, though. It's reproducing the situation where a histogram can't be expanded to include the bin associated with a recorded value, due to lack of memory. When this happens, the recorded value is accounted for in the bin with smallest upper bound that's greater than the recorded value.

Subsections `Binning` and `Update and expansion` from `docs/METRICS.md` further explain this behaviour.

That being said, this test should assert that no errors are logged as `cannot expand histogram` is just a debug entry, not an actual error.

Reply:
I considered converting the debug entry to a warning, but it worries me that such warnings could end up flooding `error.log`.